packages/react-dom/src/__tests__/ReactDOMTestSelectors-test.internal.js

```diff
@@ -0,0 +1,1427 @@
+
+      matches = findAllNodes(document.body, [
+        createComponentSelector(Outer),
+        createComponentSelector(Inner),
+        createTestNameSelector('match'),
+      ]);
+      expect(matches).toHaveLength(1);
+      expect(matches[0].id).toBe('match3');
+    });
+
+    it('should find multiple matches', () => {
+      function Example1() {
+        return (
+          <div>
+            <div data-testname="match" id="match1" />
+          </div>
+        );
+      }
+
+      function Example2() {
+        return (
+          <div>
+            <div data-testname="match" id="match2" />
+            <div data-testname="match" id="match3" />
+          </div>
+        );
+      }
+
+      render(
+        <>
+          <Example1 />
+          <Example2 />
+        </>,
+        container,
+      );
+
+      const matches = findAllNodes(document.body, [
+        createTestNameSelector('match'),
+      ]);
+      expect(matches).toHaveLength(3);
+      expect(matches.map(m => m.id).sort()).toEqual([
+        'match1',
+        'match2',
+        'match3',
+      ]);
+    });
+
+    it('should ignore nested matches', () => {
+      function Example() {
+        return (
+          <div data-testname="match" id="match1">
+            <div data-testname="match" id="match2" />
+          </div>
+        );
+      }
+
+      render(<Example />, container);
+
+      const matches = findAllNodes(document.body, [
+        createComponentSelector(Example),
+        createTestNameSelector('match'),
+      ]);
+      expect(matches).toHaveLength(1);
+      expect(matches[0].id).toEqual('match1');
+    });
+
+    it('should enforce the specific order of selectors', () => {
+      function Outer() {
+        return (
+          <>
+            <div data-testname="match" id="match1" />
+            <Inner />
+          </>
+        );
+      }
+      function Inner() {
+        return <div data-testname="match" id="match1" />;
+      }
+
+      render(<Outer />, container);
+
+      expect(
+        findAllNodes(document.body, [
+          createComponentSelector(Inner),
+          createComponentSelector(Outer),
+          createTestNameSelector('match'),
+        ]),
+      ).toHaveLength(0);
+    });
+
+    it('should not search within hidden subtrees', () => {
+      const ref1 = React.createRef(null);
+      const ref2 = React.createRef(null);
+
+      function Outer() {
+        return (
+          <>
+            <div hidden={true}>
+              <div ref={ref1} data-testname="match" />
+            </div>
+            <Inner />
+          </>
+        );
+      }
+      function Inner() {
+        return <div ref={ref2} data-testname="match" />;
+      }
+
+      render(<Outer />, container);
+
+      const matches = findAllNodes(document.body, [
+        createComponentSelector(Outer),
+        createTestNameSelector('match'),
+      ]);
+
+      expect(matches).toHaveLength(1);
+      expect(matches[0]).toBe(ref2.current);
```

Comment:
Is there a specific reason why sometimes you compare against `id` attributes and sometimes against host instances?

Reply:
just how I wrote the tests.